File: frontend/src/app/modules/user/user.service.ts
/*
 * Copyright 2012-2018 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {User} from '../../shared/model/dto/user/User';

import {Observable} from 'rxjs/Observable';
import {AbstractHttpService} from '../../shared/services/abstract-http.service';
import {map} from 'rxjs/operators';
import {Subject} from 'rxjs/Subject';
import {BehaviorSubject} from 'rxjs/BehaviorSubject';
import {TokenService} from '../../shared/auth/token.service';
import {RoleEnum} from '../../shared/model/dto/enums/RoleEnum';
import {UserSetting} from '../../shared/model/dto/UserSetting';
import {ThemeService} from '../../shared/services/theme.service';

/**
 * User service that manage users
 */
@Injectable()
export class UserService extends AbstractHttpService {

  /**
   * The base url for user API
   * @type {string} The user API url
   */
  public static readonly USERS_BASE_URL = `${AbstractHttpService.BASE_API_URL}/${AbstractHttpService.USERS_URL}`;

  /**
   * The connected user subject
   *
   * @type {Subject<User>}
   */
  connectedUserSubject: BehaviorSubject<User> = new BehaviorSubject<User>(null);

  /**
   * The constructor
   *
   * @param {HttpClient} http The http client service to inject
   * @param {TokenService} _tokenService The token service
   * @param {ThemeService} _themeService The theme service
   */
  constructor(private http: HttpClient,
              private _tokenService: TokenService,
              private _themeService: ThemeService) {
    super();
  }

  /* *************************************************************************************** */
  /*                            HTTP Functions                                               */

  /* *************************************************************************************** */

  /**
   * Get the list of users
   *
   * @returns {Observable<User[]>} The list of users
   */
  getAll(): Observable<User[]> {
    return this.http.get<User[]>(`${UserService.USERS_BASE_URL}`);
  }

  /**
   * Get a user by id
   *
   * @param {string} userId The user id to find
   * @returns {Observable<User>} The user found
   */
  getById(userId: string): Observable<User> {
    return this.http.get<User>(`${UserService.USERS_BASE_URL}/${userId}`);
  }

  /**
   * Get the connected user
   *
   * @returns {Observable<User>} The connected user
   */
  getConnectedUser(): Observable<User> {
    return this.http.get<User>(`${UserService.USERS_BASE_URL}/current`).pipe(
        map(user => {
          this.connectedUserSubject.next(user);
          return user;
        })
    );
  }

  /**
   * Delete a user
   * @param {User} user The user to delete
   */
  deleteUser(user: User): Observable<User> {
    return this.http.delete<User>(`${UserService.USERS_BASE_URL}/${user.id}`);
  }

  /**
   * Update a user
   *
   * @param {User} user The user to update
   * @returns {Observable<User>} The user updated
   */
  updateUser(user: User): Observable<User> {
    return this
        .http
        .put<User>(`${UserService.USERS_BASE_URL}/${user.id}`, user)
        .pipe(
            map(userUpdated => {
              if (userUpdated.id === this.connectedUserSubject.getValue().id) {
                this.connectedUserSubject.next(userUpdated);
              }
              return userUpdated;
            })
        );
  }

  /**
   * Update user settings
   *
   * @param {User} user The user to update
   * @param {UserSetting[]} userSettings The new settings
   * @returns {Observable<User>} The user updated
   */
  updateUserSettings(user: User, userSettings: UserSetting[]): Observable<User> {
    const url = `${UserService.USERS_BASE_URL}/${user.id}/settings`;

    return this
        .http
        .put<User>(url, userSettings)
        .pipe(
            map(userUpdated => {
              if (userUpdated.id === this.connectedUserSubject.getValue().id) {
                this.connectedUserSubject.next(userUpdated);
              }
              return userUpdated;
            })
        );
  }

  /**
   * Search a list of users by username
   *
   * @param {string} username The username to find
   * @returns {Observable<User[]>} The list of users that match the string
   */
  searchUserByUsername(username: string): Observable<User[]> {
    return this.http.get<User[]>(`${UserService.USERS_BASE_URL}/search?username=${username}`);
  }

  /* *************************************************************************************** */
  /*                            Other Help Functions                                         */

  /* *************************************************************************************** */

  /**
   * Get the initials of a user
   *
   * @param {User} user The user
   * @returns {string} The initials
   */
  getUserInitial(user: User): string {
    return `${user.firstname.substring(0, 1)}${user.lastname.substring(0, 1)}`;
  }

  /**
   * Test if the user is admin
   * @returns {boolean} True if the current user admin, false otherwise
   */
  isAdmin(): boolean {
    return this._tokenService.getUserRoles().includes(RoleEnum.ROLE_ADMIN);
  }

  /**
   * Get the template user setting
   *
   * @param {User} user The user
   * @returns {UserSetting} The user setting
   */
  getThemeUserSetting(user: User): UserSetting {
    return user.userSettings.find(userSetting => {
      return userSetting.setting.description.localeCompare('template', 'en', {sensitivity: 'base'}) === 0;
    });
  }

  /**
   * Set every user settings
   * @param {User} user The user use for set the settings
   */
  setUserSettings(user: User) {
    const userSetting = this.getThemeUserSetting(user).settingValue.value;
    this._themeService.setTheme(userSetting);
  }
}

File: src/main/java/io/suricate/monitoring/services/api/UserService.java
/*
 * Copyright 2012-2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.suricate.monitoring.services.api;

import io.suricate.monitoring.model.entities.Role;
import io.suricate.monitoring.model.entities.User;
import io.suricate.monitoring.model.enums.AuthenticationProvider;
import io.suricate.monitoring.model.enums.UserRoleEnum;
import io.suricate.monitoring.repositories.UserRepository;
import io.suricate.monitoring.services.mapper.UserMapper;
import io.suricate.monitoring.services.specifications.UserSearchSpecification;
import io.suricate.monitoring.utils.exceptions.ObjectNotFoundException;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Service used to manage user
 */
@Service
public class UserService {
    /**
     * Logger
     */
    private static final Logger LOGGER = LoggerFactory.getLogger(UserService.class);

    /**
     * The user repository
     */
    @Autowired
    private UserRepository userRepository;

    /**
     * The user mapper
     */
    @Autowired
    private UserMapper userMapper;

    /**
     * The role service
     */
    @Autowired
    private RoleService roleService;

    /**
     * The project service
     */
    @Autowired
    private ProjectService projectService;

    /**
     * The user setting service
     */
    @Autowired
    private UserSettingService userSettingService;

    /**
     * Register a new user in database authentication mode
     * @param user User to register
     * @return The user registered
     */
    public User create(User user) {
        UserRoleEnum roleEnum = userRepository.count() > 0 ? UserRoleEnum.ROLE_USER : UserRoleEnum.ROLE_ADMIN;
        Optional<Role> role = roleService.getRoleByName(roleEnum.name());
        if (!role.isPresent()) {
            LOGGER.error("Role {} not available in database", roleEnum);
            throw new ObjectNotFoundException(Role.class, roleEnum);
        }

        user.setRoles(Collections.singleton(role.get()));
        userRepository.save(user);

        user.getUserSettings().addAll(userSettingService.createDefaultSettingsForUser(user));

        return user;
    }

    /**
     * Register a new user in ldap/oauth2 authentication mode
     * @param username The username
     * @param firstname The user firstname
     * @param lastname The user lastname
     * @param email The user email
     * @param avatarUrl The user avatar URL
     * @param authenticationMethod The ID provider used
     * @return The registered user
     */
    @Transactional
    public User registerUser(String username, String firstname, String lastname, String email, String avatarUrl,
                             AuthenticationProvider authenticationMethod) {
        Optional<User> optionalUser = getOneByEmail(email);

        if (!optionalUser.isPresent()) {
            int countUsername = 1;
            while (existsByUsername(username)) {
                username = username + countUsername;
            }

            User user = userMapper.connectedUserToUserEntity(username, firstname, lastname, email, avatarUrl, authenticationMethod);
            return create(user);
        }

        return update(optionalUser.get(), optionalUser.get().getUsername(), firstname, lastname, email, avatarUrl, authenticationMethod);
    }

    /**
     * Update the user information
     * @param user                 The user to update
     * @param username             The username
     * @param firstname            The user firstname
     * @param lastname             The user lastname
     * @param email                The user email
     * @param avatarUrl            The user avatar url
     * @param authenticationMethod The ID provider used
     * @return The updated user
     */
    public User update(final User user, String username, String firstname, String lastname, String email, String avatarUrl, AuthenticationProvider authenticationMethod) {
        User userUpdated = userMapper.connectedUserToUserEntity(username, firstname, lastname, email, avatarUrl, authenticationMethod);
        userUpdated.setRoles(user.getRoles());
        userUpdated.setProjects(user.getProjects());
        userUpdated.setUserSettings(user.getUserSettings());
        userUpdated.setId(user.getId());

        userRepository.save(userUpdated);
        return userUpdated;
    }

    /**
     * Get a user by id
     *
     * @param userId The user id
     * @return The user as optional
     */
    public Optional<User> getOne(Long userId) {
        return userRepository.findById(userId);
    }

    /**
     * Get a user by username ignoring case
     * @param username The username
     * @return The user as optional
     */
    @Transactional(readOnly = true)
    public Optional<User> getOneByUsername(String username) {
        return userRepository.findByUsernameIgnoreCase(username);
    }

    /**
     * Get a user by email ignoring case
     * @param email The email
     * @return The user as optional
     */
    @Transactional(readOnly = true)
    public Optional<User> getOneByEmail(String email) {
        return userRepository.findByEmailIgnoreCase(email);
    }

    /**
     * Check if a given username exists
     * @param username The username
     * @return true if it is, false otherwise
     */
    @Transactional(readOnly = true)
    public boolean existsByUsername(String username) {
        return userRepository.existsByUsername(username);
    }

    /**
     * Get all paginated users
     *
     * @param search The specification to apply
     * @param pageable The pageable to apply
     * @return The paginated users
     */
    @Transactional
    public Page<User> getAll(String search, Pageable pageable) {
        return userRepository.findAll(new UserSearchSpecification(search), pageable);
    }

    /**
     * Delete a user
     *
     * @param user the user to delete
     */
    @Transactional
    public void deleteUserByUserId(User user) {
        user.getProjects().forEach(project -> projectService.deleteUserFromProject(user, project));
        userRepository.delete(user);
    }

    /**
     * Update a user
     *
     * @param userId    The user id
     * @param username  The username to update
     * @param firstname The firstname to update
     * @param lastname  The lastname to update
     * @param email     The email to update
     * @param roleNames The list of role names for the user
     * @return The user updated
     */
    public Optional<User> updateUser(final Long userId,
                                     final String username,
                                     final String firstname,
                                     final String lastname,
                                     final String email,
                                     final List<UserRoleEnum> roleNames) {
        Optional<User> userOpt = getOne(userId);

        if (!userOpt.isPresent()) {
            return Optional.empty();
        }

        User user = userOpt.get();

        if (StringUtils.isNotBlank(StringUtils.trimToEmpty(username))) {
            user.setUsername(username.trim());
        }

        if (StringUtils.isNotBlank(StringUtils.trimToEmpty(firstname))) {
            user.setFirstname(firstname.trim());
        }

        if (StringUtils.isNotBlank(StringUtils.trimToEmpty(lastname))) {
            user.setLastname(lastname.trim());
        }

        if (StringUtils.isNotBlank(StringUtils.trimToEmpty(email))) {
            user.setEmail(email.trim());
        }

        if (roleNames != null && !roleNames.isEmpty()) {
            this.updateUserRoles(user, roleNames);
        }

        userRepository.save(user);

        return Optional.of(user);
    }

    /**
     * Update the roles for a user
     *
     * @param user      The user
     * @param roleNames The roles to set
     */
    private void updateUserRoles(User user, List<UserRoleEnum> roleNames) {
        Set<Role> rolesToSet = roleNames.stream()
            .map(roleName -> roleService.getRoleByName(roleName.name()).orElse(null))
            .collect(Collectors.toSet());

        if (!rolesToSet.isEmpty()) {
            user.setRoles(rolesToSet);
        }
    }
}
